### PR TITLE
ENH: Remove unnecessary list collection

### PR DESCRIPTION
**Pull request by EarlMilktea**
This PR removes unnecessary list comprehensions of the form `sum([...])` or similar.

### CODE REVIEW

**Comment by seberg** [comment]
I expect most or all of these are intentional micro-optimizations because list comprehensions are fast:
```
In [4]: %timeit sum(i for i in range(10))
369 ns ± 0.798 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)

In [5]: %timeit sum([i for i in range(10)])
287 ns ± 1.59 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)
```

**Comment by EarlMilktea** [comment]
> intentional micro-optimizations

I don't think so because there are very similar genexps near listcomps, for example:

```python
N2 = math.prod(bs[axis] for axis in axes_b)               # <- genexp
newshape_a = (math.prod([as_[ax] for ax in notin]), N2)   # <- listcomp
```

I suspect they're ancient artifacts or just written as-is (no implications behind) and suggest coherently replacing them with modern genexps.

**Comment by eendebakpt** [comment]
I agree with seberg that at least some of the list comprehensions are intentional (for performance reasons). Performance with generators will not be much worse (or just as fast), but it could matter. The readability of the code with either list comprehensions or generators is comparable (in my opinion).

So I am inclined to keep the code as is, unless benchmarks show the performance is unaffected.

**Comment by EarlMilktea** [comment]
@eendebakpt 

I agree we should not touch them if performance becomes worse.

> unless benchmarks show the performance is unaffected

Could you tell me the fair way to show this? Are there built-in benchmark checks in the CI or similar?

**Comment by eendebakpt** [comment]
There are benchmarks, but I think you have to run the locally. See https://numpy.org/devdocs/benchmarking.html

Another option is to create dedicated (micro)benchmarks. For example for `np.tensordot` (one of the methods changed in this PR) a simple benchmark is given in https://github.com/numpy/numpy/pull/25788. For that benchmark `np.tensordor` has a running time of about 14 us. With the changes about 160 ns would be added (based on the microbenchmark from seberg), so that might be acceptable. (Note: that is for an 64 x 64 array, for smaller arrays the overhead could be larger)

Also note that the changes in this PR are at places where the list / generator will never have a large size (the iteration is typically over the number of input arrays)

**Comment by seberg** [comment]
To be honest, I am just not sure I care either way.  It's unlikely to be a visible micro-optimization at least in the majority of cases, but I also don't think it is such a big improvement.
(The distutils code is also barely maintained.)

So if anyone else cares either way, I'll give my +1 to just go ahead ;).

> There are benchmarks, but I think you have to run the locally

Yeah, I would be surprised many of relevant benchmarks, currently.

> listcomp is actually faster than I expected even for fairy large n

Yes and we are really interested in very small n here.

File: numpy/_core/numeric.py [hunk 1]
```diff
@@ -1163,13 +1163,13 @@ def tensordot(a, b, axes=2):
...
     N2 = math.prod(as_[axis] for axis in axes_a)
-    newshape_a = (math.prod([as_[ax] for ax in notin]), N2)
+    newshape_a = (math.prod(as_[ax] for ax in notin), N2)
     olda = [as_[axis] for axis in notin]
 
     notin = [k for k in range(ndb) if k not in axes_b]
     newaxes_b = axes_b + notin
     N2 = math.prod(bs[axis] for axis in axes_b)
-    newshape_b = (N2, math.prod([bs[ax] for ax in notin]))
+    newshape_b = (N2, math.prod(bs[ax] for ax in notin))
```
Comment (ngoldbaum) [inline numpy/_core/numeric.py:1172]: These are the only changes that *may* be possibly impactful on performance, and I agree that it shouldn't matter over other stuff that's happening in this function. The other use outside of distutils and tests is in a `__repr__` implementation, and printing performance isn't a big deal either.

IMO this change is fine and is marginally more readable.

**Review by ngoldbaum — COMMENTED** [review]
We don't maintain `numpy.distutils` anymore - can you revert the changes to that module? It will be deleted in a few years when Python 3.12 becomes the minimum supported version.

Everything else is fine, but it would be nice to turn on a lint for this (seems like the sort of thing ruff has a check for), and explicitly ignore `numpy.distutils` in the linter.

**Comment by EarlMilktea** [comment]
> revert the changes to that module

Done.

> turn on a lint for this

That's the point of this PR: always using genexp is maintainer-friendly and may be preferable in general.
I came to understand listcomp is slightly faster than genexp (thank you @seberg , I didn't know that) but for long iterators genexp is still preferred for memory optimization, and I personally believe that merits of the latter surpluses the former, especially when the project is so big (`numpy` is big project, isn't it?) that maintainers cannot examine every single line of the codes and perform microbench for each iterator reduction.
By the way if you maintainers are interested in `ruff` integration, I'm willing to work on that.

File: numpy/_core/numeric.py (hunk 1 above)
Comment (EarlMilktea) [reply numpy/_core/numeric.py:1172]: How do you think about https://github.com/numpy/numpy/pull/27884#issuecomment-2515346047 ?

**Comment by eendebakpt** [comment]
Some work with ruff has already been done, see https://github.com/numpy/numpy/issues/24994. Not sure what the status is though, so perhaps a helping hand would be nice.

**Pull request merged**